### PR TITLE
Userland: Speed up the execution of the cut command.

### PR DESCRIPTION
**Pull request by marprok**
Avoid creating SingleIndexes in case of byte ranges. This, boosts the performance
significantly in case a byte range is too big(e.g 666-123123).
Also, claim copyright over this mess since I am the one responsible for it.

### CODE REVIEW

File: Userland/cut.cpp [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
- * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2018-2020, Marios Prokopakis <mariosprokopakis@gmail.com>
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are met:
  *
- * 1. Redistributions of source code must retain the above copyright notice, this
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this
```
Comment (awesomekling) [inline Userland/cut.cpp]: Please don't tweak the copyright headers.

File: Userland/cut.cpp [hunk 2]
```diff
@@ -14,7 +15,8 @@
  *
  * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
- * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE
```
Comment (awesomekling) [inline Userland/cut.cpp]: Please don't tweak the copyright headers.

File: Userland/cut.cpp [hunk 3]
```diff
@@ -34,11 +36,20 @@
...
     ssize_t m_from { -1 };
     ssize_t m_to { -1 };
     Type m_type { Type::SingleIndex };
+
+    bool intersects(const Index& other)
+    {
+        if (m_type != Type::RangedIndex)
+            return m_from == other.m_from;
+        else
+            return !(other.m_from > m_to || other.m_to < m_from);
```
Comment (awesomekling) [inline Userland/cut.cpp]: No need for `else` after `return`.

File: Userland/cut.cpp [hunk 4]
```diff
@@ -49,8 +60,20 @@ static void print_usage_and_exit(int ret)
 
 static void add_if_not_exists(Vector<Index>& indexes, Index data)
 {
-    auto find = [data](auto& other) { return other.m_from == data.m_from && other.m_to == data.m_to; };
-    if (indexes.find(find) == indexes.end()) {
+    auto min = [](ssize_t a, ssize_t b) { return a < b ? a : b; };
+    auto max = [](ssize_t a, ssize_t b) { return a > b ? a : b; };
```
Comment (awesomekling) [inline Userland/cut.cpp]: We already have `min()` and `max()` in `<AK/StdLibExtras.h>`, please use them instead of defining custom ones :)

File: Userland/cut.cpp [hunk 5]
```diff
@@ -64,15 +87,17 @@ static void expand_list(Vector<String>& tokens, Vector<Index>& indexes)
         }
 
         if (token == "-") {
-            fprintf(stderr, "cut: invalid range with no endpoint: %s\n", token.characters());
+            fprintf(stderr, "cut: invalid range with no endpoint: %s\n",
+                token.characters());
```
Comment (awesomekling) [inline Userland/cut.cpp]: There seems to be a lot of unrelated formatting changes like this. Are you running an old `clang-format` perhaps?

Comment (marprok) [reply Userland/cut.cpp]: Maybe, I did ran it and it messed with my enum. I didn't do anything by hand.

File: Userland/cut.cpp (hunk 4 above)
Comment (marprok) [reply Userland/cut.cpp]: The AK library is so dense that is getting harder and harder to see what exists and what not. Maybe a small doxygen like documentation would help :P

File: Userland/cut.cpp (hunk 2 above)
Comment (marprok) [reply Userland/cut.cpp]: clang-format did that and I don't know why. I will rerun it.

**Comment by marprok** [comment]
@awesomekling The problem was with the version of clang-format I was using. Once again I thank you for your patience :)

**Comment by awesomekling** [comment]
Okay, this is a lot nicer :)
